<h1 style="text-align:center">Выплаты по кредиту. Решение трансцендентного уравнения</h1>

<p style="text-align:center">Нет входных данных<br>
Нет выходных данных<br>
Лимит по времени: 100 с<br>
Лимит по памяти: 128 МБ<br></p>

<h2>Условие</h2>
<p>Предположим, что кто-то желает взять кредит на сумму $100000. Он выбирает среди следующих возможностей:</p>

<ol style="margin-left:10%">
    <li>взять кредит на 10 лет с ежемесячными выплатами в $1500;</li>
    <li>взять кредит на 15 лет с ежемесячными выплатами в $1250;</li>
    <li>взять кредит на 20 лет с ежемесячными выплатами в $1160.</li>
</ol>

<p>Какой кредит выгоднее брать?<br>
Можно просто вычислить переплату, а можно еще учесть процентные ставки.<br>
Сами процентные ставки здесь не указаны, но можно их найти из основного финансового уравнения:</p>

$$
{\rm PV}\, (1+i)^n + {\rm PMT}\, (1+i)^{n-1} + {\rm PMT}\, (1+i)^{n-2} + \dots + {\rm PMT} + {\rm FV} = 0, \tag{1}
$$

<p>связывающего следующие величины:</p>

<ul style="margin-left:10%">
    <li>PV - величина кредита (present value);</li>
    <li>FV - баланс после n платежей (future value);</li>
    <li>PMT - величина периодического платежа со знаком минус (payment);</li>
    <li>n - количество платежей;</li>
    <li>i - процентная ставка (interest rate).</li>
</ul>

<p>Например, для случая 1):</p>

<ul style="margin-left:10%">
    <li>PV = 100000;</li>
    <li>FV = 0 (после всех платежей банк и частное лицо друг другу ничего не должны);</li>
    <li>PMT = −1500;</li>
    <li>n = 10×12 = 120.</li>
</ul>

<p>Используя формулу для суммы элементов геометрической прогрессии, уравнение (1) запишем в следующем виде:</p>

$$
{\rm PV}\, (1+i)^n + {\rm PMT}\, \frac{(1+i)^n - 1}{i} + {\rm FV} = 0, \tag{2}
$$

<p>Относительно i это уравнение в общем случае аналитические решено быть не может, поэтому надо использовать численные методы.</p>

<ol style="margin-left:10%">
    <li>Написать функцию, решающее уравнение (2). Ваша функция должна вызывать scipy.optimize.brentq. Можете воспользоваться другим методом. Обоснуйте свой выбор.</li>
    <li>Рассчитать годовые процентные ставки для примеров 1)-3). Годовая процентная ставка равна i (year) = mi, где m - количество платежей в год (в нашем случае m = 12).</li>
    <li>Сравнить процентные ставки. Какой кредитный план выбрали бы вы?</li>
</ol>

<p><b>Замечание.</b> Уравнения (1), (2) можно использовать и для рассчета выплат по вкладам. В этом случае PV < 0, PMT ≥ 0.</p>

In [23]:
import numpy as np
import scipy as sp
from scipy.optimize import brentq

In [24]:
pv = 100000
fv = 0
pmt1 = -1500
n1 = 10 * 12
pmt2 = -1250
n2 = 15 * 12
pmt3 = -1160
n3 = 20 * 12

In [25]:
def sec_equation_first(i):
    return pv * (1 + i) ** n1 + pmt1 * (((1 + i) ** n1 - 1) / i) + fv 
def sec_equation_second(i):
    return pv * (1 + i) ** n2 + pmt2 * (((1 + i) ** n2 - 1) / i) + fv 
def sec_equation_third(i):
    return pv * (1 + i) ** n3 + pmt3 * (((1 + i) ** n3 - 1) / i) + fv

def def_solve(pv, pmt, n, fv):
    def fun(i):
        return pv * (1 + i) ** n + pmt * ((((1 + i) ** n) - 1) / i) + fv
    return brentq(fun, 0.001, 1.0)

In [26]:
solution_one = brentq(sec_equation_first, 0.01, 1.0)
solution_second = brentq(sec_equation_second, 0.01, 1.0)
solution_third = brentq(sec_equation_third, 0.01, 1.0)
print(f'1 solutin: {solution_one}')
print(f'2 solution: {solution_second}')
print(f'3 solution: {solution_third}')

1 solutin: 0.010930603939868333
2 solution: 0.01063980717924384
3 solution: 0.010697698792848072


In [29]:
month_procent1 = def_solve(pv, pmt1, n1, fv)
month_procent2 = def_solve(pv, pmt2, n2, fv)
month_procent3 = def_solve(pv, pmt3, n3, fv)
print(f'1. Ежемесячный процент: {month_procent1}')
print(f'2. Ежемесячный процент: {month_procent2}')
print(f'3. Ежемесячный процент: {month_procent3}')

1. Месячный процент: 0.010930603939676818
2. Месячный процент: 0.01063980717901046
3. Месячный процент: 0.010697698792974464


In [28]:
year_procent1 = 12 * month_procent1
year_procent2 = 12 * month_procent2
year_procent3 = 12 * month_procent3
print(f'Кредит на 10 лет, годовая ставка = {year_procent1 * 100}%')
print(f'Кредит на 15 лет, годовая ставка = {year_procent2 * 100}%')
print(f'Кредит на 20 лет, годовая ставка = {year_procent3 * 100}%')

Кредит на 10 лет, годовая ставка = 13.11672472761218%
Кредит на 15 лет, годовая ставка = 12.767768614812553%
Кредит на 20 лет, годовая ставка = 12.837238551569358%


## Вывод

Годовые ставки для кредита на сумму $\$100000$. составили:
* $13.12 \% $ для варианта 1
* $12.77 \% $ для варианта 2
* $12.84 \% $ для варианта 3

Из данных вариантов наиболее выгодным в плане годовой ставки является 2 вариант.